### PR TITLE
[TRUNK-15323] Recurse on paths that look file-ish

### PR DESCRIPTION
**Pull request by cmillar-trunk**
If a path ends without a slash, github will recurse into it for codeowners, so we add a fallback path to recurse in any directory that matches, as well as matching on any files that match.

### CODE REVIEW

**Review by gnalh — APPROVED** [review]
This looks reasonable to me and thoroughly tests the various cases we can hit. Any concerns with this approach @dfrankland ?

**Comment by dfrankland** [comment]
I'm going to check this branch out and see if there's an edge case we're missing

File: codeowners/src/github.rs [hunk 1]
```diff
@@ -187,7 +185,54 @@ impl BindingsGitHubOwners {
...
+pub struct PatternWithFallback {
+    base: Pattern,
+    fallback: Option<Pattern>,
+}
+
+impl PatternWithFallback {
+    pub fn new(base: &str) -> anyhow::Result<Self> {
+        // Matches anything that ends with neither a slash nor a period nor an asterisk,
+        // see test pattern_with_fallback for cases
+        let fallback_needed_regex = Regex::new(r"^\/?([a-zA-Z0-9_*\-]*\/)*([a-zA-Z0-9_\-]+)$")?;
```
Comment (dfrankland) [inline codeowners/src/github.rs:198]: This regex doesn't work for hidden paths which is something that is used by customers. For example this fails:

```rs
assert_eq!(
    PatternWithFallback::new("/.buildkite").unwrap(),
    PatternWithFallback {
        base: Pattern::new("/.buildkite").unwrap(),
        fallback: Some(Pattern::new("/.buildkite/**").unwrap()),
    },
);
```

Additionally, this won't support any paths with a period in it. Although this might be an edge-case we don't support, I think we should document that we don't support it yet. We can do that through a test or otherwise. For example this fails:

```rs
assert_eq!(
    PatternWithFallback::new("/top.level/second-level").unwrap(),
    PatternWithFallback {
        base: Pattern::new("/top.level/second-level").unwrap(),
        fallback: Some(Pattern::new("/top.level/second-level/**").unwrap()),
    },
);
```

Comment (dfrankland) [inline codeowners/src/github.rs:198]: We should use `lazy_static` for this regex to avoid recompiling it on every invocation. Docs:
https://docs.rs/regex/latest/regex/#avoid-re-compiling-regexes-especially-in-a-loop

Comment (cmillar-trunk) [reply codeowners/src/github.rs:198]: Nice catch. I added logic to support those paths and that extraction.

**Review by dfrankland — APPROVED** [review]

File: codeowners/src/github.rs [hunk 2]
```diff
@@ -187,7 +186,55 @@ impl BindingsGitHubOwners {
...
+    base: Pattern,
+    fallback: Option<Pattern>,
+}
+
+impl PatternWithFallback {
+    pub fn new(base: &str) -> anyhow::Result<Self> {
+        // Matches anything that ends with neither a slash nor a period nor an asterisk,
+        // see test pattern_with_fallback for cases
+        static FALLBACK_NEEDED_REGEX: Lazy<Regex> =
+            Lazy::new(|| Regex::new(r"^\/?(.*\/)*((\.?)[^\/\.\*]+)$").unwrap());
```
Comment (dfrankland) [inline codeowners/src/github.rs:200]: This looks fine. In my humble opinion, I'd simplify this to just looking for `/\*{0,2}$` (look for a literal `/` character plus zero to two literal `*` characters at the end of the string):
https://regex101.com/r/gFPdbL/1

Then I would check for `if !FALLBACK_NEEDED_REGEX.is_match(base) { ... }` below.

That's up to you if you like that or not.

**Review by dfrankland — APPROVED** [review]

File: codeowners/src/github.rs (hunk 2 above)
Comment (cmillar-trunk) [reply codeowners/src/github.rs:200]: I like the simplicity, but I'd like to avoid trying to recurse into files with extensions, which requires a more complex pattern.
